### ISSUE TITLE
Creating new line inside H-blocks

### ISSUE THREAD

**Issue by Greeny7** [yabwe/medium-editor#1060]
There are some problems with headings - they are dissapearing when you make new line with caret inside this block.
I think it is about the same problem, which is this issue/PR about - https://github.com/yabwe/medium-editor/pull/382
Now creating new line in headings/block containers works good, but only if we have no nested tags inside. For example, if you make your H-tag underlined (by creating <u> tag inside), and hit return inside this block - function handleKeyup() will get <u> as parent block and will not prevent old editor behavior.
![editor-bug](https://cloud.githubusercontent.com/assets/12188900/14533637/8a8387b2-026e-11e6-99a3-b194cda00788.gif)

**Comment by orthes** [comment]
Thanks for reporting this. I tried this in a pure `contenteditable` element without medium-editor to see if it's something that's broken in `contenteditable` itself, or we're doing something wrong there. And it looks like `contenteditable` without medium-editor works correctly, so we're screwing something up here.

**Comment by Greeny7** [comment]
@orthes Yes, if I understand right, Medium Editor creating P elements on return-key hit to prevent the creation of div's - because div element creation on new line is a default behavior of condenteditable.
It follow from the handleKeyup() function in core.js. 
And there is a condition, that on return key hit it must check the type of parent tag and if it is a block element, like h1/h2/quote etc. - it must prevent formatBlock execution on new line.

I don't fully understand yet in what situations we need to prevent div-creating and how we must build the condition in handleKeyup() function, so I made a temporary repair only for H-tags (because headings destroying is most noticeable and ugly):
https://github.com/Greeny7/medium-editor/commit/302a0ecfbb1cd5a33c81d1e5577ae4c2759b8c06

I would be grateful if someone will clarify how to do a better condition to fully resolve this issue.

**Comment by nmielnik** [comment]
@Greeny7 this is definitely something that needs to be understood and addressed.  I think the commit you pointed to makes sense, and it could be worth opening a PR agains the main repo and see if we can try to add some tests and improve upon it more iteratively to see if any other issues might come out of it.

There are many issues open related to the chunk of code that handles `keyup` in core.js (#907, #834, #829, #865, #149, and more).

There have also been a few PRs recently making some adjustments to this code ( #997, #994, #986).

It doesn't seem like any one person specifically has a good grasp of this code and what all it's intended to cover.  However, due to how critical that code path is, and all the problems that keep coming up, we might just have to keep experimenting with fixes like what you proposed and see what happens 😕
